flow/designs/asap7/mock-array/Element/constraints.sdc

```diff
@@ -4,34 +4,20 @@ set cols [expr {[info exists ::env(MOCK_ARRAY_COLS)] ? $::env(MOCK_ARRAY_COLS) :
 
 set clk_name clock
 set clk_port_name clock
-set clk_period 1000
+set clk_period 500
 
 set clk_port [get_ports $clk_port_name]
 create_clock -period $clk_period -waveform [list 0 [expr $clk_period / 2]] -name $clk_name $clk_port
 set_clock_uncertainty -setup 20.0 [get_clocks $clk_name]
 set_clock_uncertainty -hold 20.0 [get_clocks $clk_name]
 
+set clock_network_latency 100
+
 # io_ins_x -> REG_x in neighbouring element or just outside of the array
-set_input_delay -clock $clk_name [expr $clk_period * 0.05] [get_ports {io_ins_*}]
+set_input_delay -clock $clk_name [expr $clock_network_latency + $clk_period * 0.05] [get_ports {io_ins_* io_lsbIns_4}]
```

Comment:
Why just io_lsbIns_4 ?

Reply:
will remove, belongs to a different PR

Reply:
Is the question about what if clock latency greater than clock period and how to constraints IO paths accordingly?
One way to compensate for "actual clock tree delay/latency" is the use of virtual clock to constraint the IO.
(It has similar effect as suggestion from @rovinski, but I guess his is using real clock to constraint the IO)

Reply:
> Is the question about what if clock latency greater than clock period and how to constraints IO paths accordingly?

Not really, I think that is a separate concern that I did not intend to address in this PR.

My thinking is that clock can be inserted *earlier* when mock-array is used as part of a macro where the clock network latency is significantly greater than the clock network latency of mock-array as long as the clock tree of the macro that contains mock-array is balanced approperiately.

This is what I'm trying to model here and this allows me to reduce the clock period of the mock-array as well as getting rid of all the hold cells.


> One way to compensate for "actual clock tree delay/latency" is the use of virtual clock to constraint the IO. (It has similar effect as suggestion from @rovinski, but I guess his is using real clock to constraint the IO)

Is there, at this point, a reason to prefer a different approach to the one used here?

I have never used virtual clocks so I don't know how to articulate that alternative solution. Perhaps that could be a future refinement on the approach?

Reply:
@mynnguyen I've asked for a second review to follow up on the above.